src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -524,6 +550,35 @@ public override int GetCount(bool onlyIfCheap)
 
                 return !onlyIfCheap || _source is ICollection<TSource> ? _source.Count() + (_appended == null ? 0 : _appended.Count) + (_prepended == null ? 0 : _prepended.Count) : -1;
             }
+
+            public override HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer)
+            {
+                int count = GetCount(onlyIfCheap: true);
+
+                HashSet<TSource> hashSet = count == -1 ? new HashSet<TSource>(comparer) : new HashSet<TSource>(count, comparer);
```

Comment:
Very much agreed. @TiagoSoczek, it's best to just avoid the whole `GetCount` business entirely and just start with `new HashSet<TSource>(comparer)`.